extras/axelar-token/src/IpcTokenHandler.sol

```diff
@@ -78,38 +104,28 @@ contract IpcTokenHandler is InterchainTokenExecutable, IpcHandler {
 
         ResultMsg memory result = abi.decode(envelope.message, (ResultMsg));
         if (result.outcome != OutcomeType.Ok) {
-            // Note: IPC only supports deploying subnets via our blessed registry, so we can trust the code behind
-            // the subnet actor.
+            // Verify that the subnet is indeed an ERC20 subnet.
             SupplySource memory supplySource = SubnetActor(envelope.from.subnetId.getAddress()).supplySource();
             require(supplySource.kind == SupplyKind.ERC20, "expected ERC20 supply source");
 
+            // Increase the allowance of the admin address so they can retrieve these otherwise lost tokens.
+            IERC20(supplySource.tokenAddress).safeIncreaseAllowance(_admin, envelope.value);
+
             // Results will carry the original beneficiary in the 'from' address.
             address beneficiary = envelope.from.rawAddress.extractEvmAddress();
 
-            // We credit the token funds to the beneficiary. The beneficiary will have to call withdraw() to pull the
-            // funds out on this network.
-            _claims[beneficiary][supplySource.tokenAddress] += envelope.value;
-
             // Emit an event.
             emit FundingFailed(envelope.from.subnetId, beneficiary, envelope.value);
         }
 
         return bytes("");
     }
 
-    // @notice Withdraws all available balance for the specified token for the sender.
-    function withdraw(address token) external {
-        uint256 available = _claims[msg.sender][token];
-        if (available == 0) {
-            revert NothingToWithdraw();
-        }
-
-        delete _claims[msg.sender][token];
-        IERC20(token).safeTransfer(msg.sender, available);
+    // @notice The ultimate backstop in case the error-handling logic itself failed unexpectedly and we failed to
+    //         increase the recovery allowances of the admin address.
+    function adminTokenIncreaseAllowance(address token, uint256 amount) external {
+        require(msg.sender == _admin, "unauthorized");
+        IERC20(token).safeIncreaseAllowance(_admin, amount);
     }
```

Comment:
If this is our back pocket trump card to unroll any errors, can we get a simple test for the recovery?

Reply:
Done here: https://github.com/consensus-shipyard/ipc/pull/814/commits/f0a035aa8765517a30c9ebf1e31535522a6cc7ab